flink-ml-lib/src/main/java/org/apache/flink/ml/feature/maxabsscaler/MaxAbsScaler.java

```diff
@@ -0,0 +1,201 @@
+
+/**
+ * An Estimator which implements the MaxAbsScaler algorithm. This algorithm rescales feature values
+ * to the range [-1, 1] by dividing through the largest maximum absolute value in each feature. It
+ * does not shift/center the data and thus does not destroy any sparsity.
+ */
+public class MaxAbsScaler
+        implements Estimator<MaxAbsScaler, MaxAbsScalerModel>, MaxAbsScalerParams<MaxAbsScaler> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    public MaxAbsScaler() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public MaxAbsScalerModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        final String inputCol = getInputCol();
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        DataStream<Vector> inputData =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                (MapFunction<Row, Vector>)
+                                        value -> ((Vector) value.getField(inputCol)));
+        DataStream<Vector> maxAbsValues =
+                inputData
+                        .transform(
+                                "reduceInEachPartition",
+                                inputData.getType(),
+                                new MaxAbsReduceFunctionOperator())
+                        .transform(
+                                "reduceInFinalPartition",
+                                inputData.getType(),
+                                new MaxAbsReduceFunctionOperator())
+                        .setParallelism(1);
+        DataStream<MaxAbsScalerModelData> modelData =
+                DataStreamUtils.mapPartition(
+                        maxAbsValues,
+                        new RichMapPartitionFunction<Vector, MaxAbsScalerModelData>() {
+                            @Override
+                            public void mapPartition(
+                                    Iterable<Vector> values, Collector<MaxAbsScalerModelData> out) {
+                                DenseVector maxVector = (DenseVector) values.iterator().next();
+                                out.collect(new MaxAbsScalerModelData(maxVector));
+                            }
+                        });
+
+        MaxAbsScalerModel model =
+                new MaxAbsScalerModel().setModelData(tEnv.fromDataStream(modelData));
+        ReadWriteUtils.updateExistingParams(model, getParamMap());
+        return model;
+    }
+
+    /**
+     * A stream operator to compute the max values in each partition of the input bounded data
+     * stream.
+     */
+    private static class MaxAbsReduceFunctionOperator extends AbstractStreamOperator<Vector>
+            implements OneInputStreamOperator<Vector, Vector>, BoundedOneInput {
+        private ListState<DenseVector> maxState;
+        private DenseVector maxVector;
```

Comment:
If all input records are sparse vectors, this practice might waste much memory space. Is there any better way to improve this implementation?

Reply:
I think this case will not happen.  If all input records are sparse vectors, the maxAbsVector need not be a sparse vector. If a lot of values in maxAbsVector are zeros, which means the input data has a lot of invalid features, we need to check the input data.

Reply:
Got it. I'm OK with the current implementation.